source/lib/rocprof-sys/library/amd_smi.cpp

```diff
@@ -189,13 +189,77 @@ data::sample(uint32_t _dev_id)
     ROCPROFSYS_AMDSMI_GET(get_settings(m_dev_id).jpeg_activity,
                           amdsmi_get_gpu_metrics_info, sample_handle, &_gpu_metrics);
 
-    for(const auto& v_activity : _gpu_metrics.vcn_activity)
+    if(get_settings(m_dev_id).vcn_activity)
     {
-        if(v_activity != UINT16_MAX) m_vcn_metrics.push_back(v_activity);
+        if(gpu::is_vcn_activity_supported(_dev_id))
+        {
+            xcp_metrics_t metrics;
+            for(const auto& vcn_activity : _gpu_metrics.vcn_activity)
+            {
+                if(vcn_activity != UINT16_MAX) metrics.vcn_busy.push_back(vcn_activity);
+            }
```

Comment:
```suggestion
    // Helper lambda to fill busy metrics from a source array
    auto fill_busy_metrics = [](auto& dest, const auto& src) {
        for(const auto& val : src)
        {
            if(val != UINT16_MAX)
                dest.push_back(val);
        }
    };
    fill_busy_metrics(metrics.vcn_busy, _gpu_metrics.vcn_activity);
```
I see that you're doing this loop in several places. You can consider a helper function / lamdba?